File: problems/linked-list/0020-valid-parentheses.py
"""#20 Valid Parentheses
The stack holds opening brackets whose matching close has not arrived.
Time: O(n) | Space: O(n)
"""


class Solution:
    def isValid(self, s):
        opening_for = {")": "(", "]": "[", "}": "{"}
        stack = []

        for character in s:
            if character in opening_for:
                if not stack or stack.pop() != opening_for[character]:
                    return False
            else:
                stack.append(character)

        return not stack

File: problems/linked-list/0020-valid-parentheses.js
/**
 * #20 Valid Parentheses
 * Pattern: stack stores opening brackets that still need a matching close.
 * Time: O(n) | Space: O(n)
 *
 * @param {string} s
 * @return {boolean}
 */
function isValid(s) {
  const openingFor = new Map([
    [')', '('],
    [']', '['],
    ['}', '{'],
  ])
  const stack = []

  for (const character of s) {
    if (openingFor.has(character)) {
      if (stack.pop() !== openingFor.get(character)) return false
    } else {
      stack.push(character)
    }
  }

  return stack.length === 0
}

module.exports = isValid
